src/azure-account.ts

```diff
@@ -783,81 +802,68 @@ async function getEnvironments(): Promise<AzureAccountEnvironment[]> {
 		}
 	}
 
+	const result = [...staticEnvironments]; // make a clone
+
 	const config = workspace.getConfiguration('azure');
-	const ppe = config.get<AzureAccountEnvironment>('ppe');
+	const ppe = config.get<Environment>('ppe');
 	if (ppe) {
-		return await getPpeEnvironments(ppe, config);
+		result.push({
+			...ppe,
+			name: azurePPE,
+			validateAuthority: getValidateAuthority(ppe.activeDirectoryEndpointUrl)
+		});
 	}
-	return staticEnvironments;
-}
 
-async function getPpeEnvironments(ppe: AzureAccountEnvironment, config: WorkspaceConfiguration): Promise<AzureAccountEnvironment[]> {
-	if (!ppe.activeDirectoryEndpointUrl) {
-		return [
-			...staticEnvironments,
-			{
-				...ppe,
-				name: azurePPE
-			}
-		];
+	const customCloudEnvironment: Environment | undefined = await getCustomCloudEnvironment(config, includePartial);
+	if (customCloudEnvironment) {
+		result.push(customCloudEnvironment);
 	}
-	
-	// get api profile from user setting, this needs to be true for running azure stack
-	const apiProfile = config.get<boolean>('target_azurestack_api_profile');
-	// get validateAuthority from activeDirectoryUrl from user setting, it should be set to false only under ADFS environemnt.
-	const activeDirectoryUrl = ppe.activeDirectoryEndpointUrl.endsWith('/') ? ppe.activeDirectoryEndpointUrl.slice(0,-1) : ppe.activeDirectoryEndpointUrl;
-	const validateAuthority = activeDirectoryUrl.endsWith('/adfs') ? false : true;
-	if (apiProfile) {
-		return await getAzureStackEnvironments(ppe, validateAuthority)
-	} else {
-		return [
-			...staticEnvironments,
-			{
-				...ppe,
-				name: azurePPE,
-				validateAuthority: validateAuthority
+
+	return result;
+}
+
+async function getCustomCloudEnvironment(config: WorkspaceConfiguration, includePartial: boolean): Promise<Environment | undefined> {
+	const armUrl = config.get<string>(customCloudArmUrlKey);
+	if (armUrl) {
+		try {
+			const endpointsUrl = getMetadataEndpoints(armUrl);
+			const endpointsResponse = await fetch(endpointsUrl);
+			if (endpointsResponse.ok) {
+				const endpoints: IResourceManagerMetadata = await endpointsResponse.json();
+				return <Environment>{
+					name: azureCustomCloud,
+					resourceManagerEndpointUrl: armUrl,
+					activeDirectoryEndpointUrl: endpoints.authentication.loginEndpoint.endsWith('/') ? endpoints.authentication.loginEndpoint : endpoints.authentication.loginEndpoint.concat('/'),
+					activeDirectoryGraphResourceId: endpoints.graphEndpoint,
+					activeDirectoryResourceId: endpoints.authentication.audiences[0],
+					portalUrl: endpoints.portalEndpoint,
+					galleryEndpointUrl: endpoints.galleryEndpoint,
+					storageEndpointSuffix: armUrl.substring(armUrl.indexOf('.')),
+					keyVaultDnsSuffix: '.vault'.concat(armUrl.substring(armUrl.indexOf('.'))),
+					managementEndpointUrl: endpoints.authentication.audiences[0],
+					validateAuthority: getValidateAuthority(endpoints.authentication.loginEndpoint)
+				};
 			}
-		]
+		} catch {
+			// ignore
```

Comment:
Let's show a notification to the user if this fails, so that they can potentially take action to fix the endpoints url